packages/table/src/CellDefinitions/ActionCell.tsx

```diff
@@ -35,12 +105,25 @@ const ActionCell = <T extends IdType>({
           ))}
         </TableActionMenu>
         {primaryAction && isPrimaryActionVisible && (
-          <Icon
-            data-testid={`table_row_action_menu_item_${id}_primaryAction`}
-            name={primaryAction.iconName}
-            title={primaryAction.title}
-            onClick={() => primaryAction.onClick(original)}
-          />
+          <>
+            <Icon
+              id={`table_row_action_menu_item_${id}_primaryAction`}
+              data-testid={`table_row_action_menu_item_${id}_primaryAction`}
+              name={
+                typeof primaryAction.iconName === 'function' ? primaryAction.iconName(original) : primaryAction.iconName
```

Comment:
Again adding support to allow either hard coding the icon that is displayed or utilizing a function and the record to determine it.